### PR TITLE
bpo-42152: Optimize out uses of PyDict_GetItemWithError if its result is not used.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Use PyDict_Contains and PyDict_SetDefault if appropriate.

https://bugs.python.org/issue42152

### CODE REVIEW

File: Python/import.c [hunk 1]
```diff
@@ -1018,14 +1018,14 @@ module_dict_for_exec(PyThreadState *tstate, PyObject *name)
...
-            _PyDict_SetItemId(d, &PyId___builtins__,
-                              PyEval_GetBuiltins()) != 0)
-        {
-            remove_module(tstate, name);
-            return NULL;
-        }
+    int r = _PyDict_ContainsId(d, &PyId___builtins__);
+    if (r == 0) {
+        r = _PyDict_SetItemId(d, &PyId___builtins__,
+                              PyEval_GetBuiltins());
```
Comment (methane) [inline Python/import.c:1024]: `PyEval_GetBuiltins()` is cheap function so we can use `PyDict_SetDefault()` here, and `builtin_eval_impl`, `builtin_exec_impl` if you want to.

**Review by methane — APPROVED** [review]
Looks very nice.

**Comment by markshannon** [comment]
LGTM.
Nice improvement in code clarity.

The PR title say this is an optimization, but I don't see that as relevant or justified.
Probably best to describe this as a code quality improvement if adding a NEWS item.

File: Python/import.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Python/import.c:1024]: There is no `_PyDict_SetDefaultId()` and I do not want to introduce it just for these three cases. Using `PyDict_SetDefault()` + `_PyUnicode_FromId()` will not make large difference:
```c
    int r = _PyDict_ContainsId(d, &PyId___builtins__);
    if (r == 0) {
        r = _PyDict_SetItemId(d, &PyId___builtins__,
                              PyEval_GetBuiltins());
    }
    if (r < 0) {
        remove_module(tstate, name);
        return NULL;
    }
```
vs
```c
    PyObject *tmp = _PyUnicode_FromId(&PyId___builtins__);
    if (tmp != NULL) {
        tmp = PyDict_SetDefault(d, tmp, PyEval_GetBuiltins());
    }
    if (tmp == NULL) {
        remove_module(tstate, name);
        return NULL;
    }
```

**Pull request merged**